File: ruby/lib/boardly/notify/email.rb
# frozen_string_literal: true

require "mail"

module Boardly
  module Notify
    # Sends reports as email over SMTP (via the `mail` gem).
    class EmailChannel
      def initialize(config)
        @config = config
        delivery = {
          address: config[:host], port: config[:port],
          enable_starttls_auto: !config[:secure],
          tls: config[:secure]
        }
        if config[:user]
          delivery[:user_name] = config[:user]
          delivery[:password] = config[:pass]
          delivery[:authentication] = :login
        end
        @delivery = delivery
      end

      def name = "email"
      def target = @config[:to].join(", ")

      def send(report)
        cfg = @config
        delivery = @delivery
        mail = Mail.new do
          from    cfg[:from]
          to      cfg[:to]
          subject report.title
          body    report.markdown
          delivery_method :smtp, delivery
        end
        mail.deliver!
      end
    end
  end
end

File: src/notify/email.ts
import nodemailer, { type Transporter } from "nodemailer";
import type { NotifyChannel, Report } from "./types.js";

export interface EmailConfig {
  host: string;
  port: number;
  secure: boolean;
  user?: string;
  pass?: string;
  from: string;
  to: string[];
}

function escapeHtml(s: string): string {
  return s.replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;");
}

/** Sends reports as email over SMTP (via nodemailer). */
export class EmailChannel implements NotifyChannel {
  readonly name = "email";
  readonly target: string;
  private readonly transport: Transporter;

  constructor(private readonly cfg: EmailConfig) {
    this.target = cfg.to.join(", ");
    this.transport = nodemailer.createTransport({
      host: cfg.host,
      port: cfg.port,
      secure: cfg.secure,
      auth: cfg.user ? { user: cfg.user, pass: cfg.pass } : undefined,
    });
  }

  async send(report: Report): Promise<void> {
    await this.transport.sendMail({
      from: this.cfg.from,
      to: this.cfg.to,
      subject: report.title,
      text: report.markdown,
      // Preserve the markdown layout in HTML-capable clients without a full renderer.
      html: `<pre style="font-family:inherit;white-space:pre-wrap">${escapeHtml(report.markdown)}</pre>`,
    });
  }
}
